crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -214,4 +208,81 @@ impl QueryLimitsChecker {
 
         Ok(())
     }
+
+    fn analyze_selection_set(
+        &self,
+        cfg: &ServiceConfig,
+        fragment_defs: &HashMap<Name, Positioned<FragmentDefinition>>,
+        sel_set: &Positioned<SelectionSet>,
+    ) -> ServerResult<ComponentCost> {
+        // Use BFS to analyze the query and
+        // count the number of nodes and the depth of the query
+
+        // Queue to store the nodes at each level
+        let mut que = VecDeque::new();
+        // Number of nodes in the query
+        let mut num_nodes: u32 = 0;
+        // Depth of the query
+        let mut depth: u32 = 0;
+        // Number of nodes at each level
+        let mut level_len;
+
+        for top_level_sel in sel_set.node.items.iter() {
+            que.push_back(top_level_sel);
+            num_nodes += 1;
+            self.check_limits(cfg, num_nodes, depth, Some(top_level_sel.pos))?;
+        }
+
+        // Track the number of nodes at first level if any
+        level_len = que.len();
+
+        while !que.is_empty() {
```

Comment:
I had some time to think about how to structure this code to avoid the `unwrap` when popping the queue, as well as relying on the invariants around levels.  I think your best bet is to use two stacks rather than a queue.  You can also clean things up slightly more by counting `num_nodes` on `pop` rather than on `push`:

```rust
let mut curr_level = vec![];
let mut next_level = vec![];

next_level.extend(sel_set.node.items.iter());

let mut num_nodes = 0;
let mut depth = 0;
while !next_level.is_empty() {
    mem::swap(&mut next_level, &mut curr_level);

    depth += 1;
    while let Some(node) = curr_level.pop() {
        num_nodes += 1;
        // I would inline `check_limits` here.

        match node {
            // as before, but without the `check_limits` call or bumping `num_nodes`.
        }
    }
}
```

Reply:
Thanks! I will address this in a subseq PR as its unrelated to the core idea here.
I'm thinking of switching to DFS altogether.

Reply:
Sounds good to me! (depth calculation is a little bit more annoying with DFS though -- elements in your stack need to remember the depth at which they were pushed on).